sklearn/metrics/_pairwise_distances_reduction/_base.pxd.tp

```diff
@@ -17,19 +17,44 @@ cimport numpy as cnp
 
 from cython cimport final
 
-from ...utils._typedefs cimport ITYPE_t, DTYPE_t
+from ...utils._typedefs cimport ITYPE_t, DTYPE_t, SPARSE_INDEX_TYPE_t
 
 cnp.import_array()
 
 {{for name_suffix, INPUT_DTYPE_t, INPUT_DTYPE in implementation_specific_values}}
 
 from ._datasets_pair cimport DatasetsPair{{name_suffix}}
 
-cpdef DTYPE_t[::1] _sqeuclidean_row_norms{{name_suffix}}(
+cpdef DTYPE_t[::1] _sqeuclidean_row_norms{{name_suffix}}_dense(
     const {{INPUT_DTYPE_t}}[:, ::1] X,
     ITYPE_t num_threads,
 )
 
+cpdef DTYPE_t[::1] _sqeuclidean_row_norms{{name_suffix}}_sparse(
+    const {{INPUT_DTYPE_t}}[:] X_data,
+    const SPARSE_INDEX_TYPE_t[:] X_indptr,
+    ITYPE_t num_threads,
+)
+
+cdef class SqeuclideanRowNorm{{name_suffix}}:
+    cdef:
+        ITYPE_t num_threads
+
+    cpdef DTYPE_t[::1] compute(self)
+
+
+cdef class DenseSqeuclideanRowNorm{{name_suffix}}(SqeuclideanRowNorm{{name_suffix}}):
+    cdef:
+        const {{INPUT_DTYPE_t}}[:, ::1] X
+
+
+cdef class SparseSqeuclideanRowNorm{{name_suffix}}(SqeuclideanRowNorm{{name_suffix}}):
+    cdef:
+        const {{INPUT_DTYPE_t}}[:] X_data
+        const SPARSE_INDEX_TYPE_t[:] X_indices
+        const SPARSE_INDEX_TYPE_t[:] X_indptr
```

Comment:
Having abstractions for this seems a bit of an overkill, but this also proves that the current computations and storing of rows' norms is misplaced.

For the scope of this PR, let's focus on making the sparse-sparse case work and refactor the row norms (removing the associated logic duplication in Euclidean specialisation) as part of another dedicated PR.